File: python/nemo_relay/memory_providers/conformance.py
# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Reusable supported-profile conformance for Python memory adapters."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from nemo_relay.memory import (
    MemoryCapabilities,
    MemoryContent,
    MemoryContractError,
    MemoryErrorCode,
    MemoryFilter,
    MemoryMaintenanceAction,
    MemoryMaintenanceRequest,
    MemoryMaintenanceResult,
    MemoryMaintenanceWindow,
    MemoryNamespace,
    MemoryProvenance,
    MemoryProvider,
    MemoryProviderError,
    MemoryRequestContext,
    MemorySearchRequest,
    MemorySearchResult,
    MemorySearchScope,
    MemoryStoreDisposition,
    MemoryStoreRequest,
)

@dataclass(frozen=True, slots=True)
class ConformanceCase:
    """Result of one named provider invariant."""

    name: str
    passed: bool
    message: str | None = None


@dataclass(frozen=True, slots=True)
class ConformanceReport:
    """Structured outcome of one isolated provider conformance run."""

    provider: str
    run_id: str
    cases: tuple[ConformanceCase, ...]

    @property
    def passed(self) -> bool:
        """Return whether every supported case passed."""
        return all(case.passed for case in self.cases)

    @property
    def failures(self) -> tuple[ConformanceCase, ...]:
        """Return only failed cases."""
        return tuple(case for case in self.cases if not case.passed)


async def run_provider_conformance(provider: MemoryProvider, run_id: str) -> ConformanceReport:
    """Run required and advertised Python adapter cases in an isolated namespace."""
    cases: list[ConformanceCase] = []
    if not run_id.strip():
        return ConformanceReport(provider.name, run_id, (ConformanceCase("valid_run_id", False, "run_id is empty"),))

    origin = _namespace(run_id, "subject", "session-a", "assistant")
    cross_session = _namespace(run_id, "subject", "session-b", "assistant")
    other_subject = _namespace(run_id, "other-subject", "session-b", "assistant")
    stored_request = _store_request(run_id, "seed", origin, "conformance solarized editor preference", "seed-key")

    try:
        stored = await provider.store(stored_request)
        _require(stored.disposition is MemoryStoreDisposition.CREATED, "initial store did not report created")
        cases.append(ConformanceCase("required_store", True))
    except Exception as error:
        cases.append(_failed("required_store", error))
        for name in (
            "required_search",
            "record_preservation",
            "subject_isolation_cross_session",
            "deterministic_order",
            "idempotency",
            "metadata_time_filters",
            "typed_invalid_request",
            "capability_agreement",
        ):
            cases.append(ConformanceCase(name, False, "required seed store failed"))
        return ConformanceReport(provider.name, run_id, tuple(cases))

    search_request = _search_request(run_id, "required-search", cross_session, "solarized editor preference")
    try:
        searched = await provider.search(search_request)
        _require(searched.matches, "search returned no matches")
        cases.append(ConformanceCase("required_search", True))
    except Exception as error:
        searched = MemorySearchResult()
        cases.append(_failed("required_search", error))

    try:
        match = next((item for item in searched.matches if item.record.id == stored.record.id), None)
        _require(match is not None, "stored record was not returned")
        assert match is not None
        _require(match.record == stored.record, "stored record fields changed")
        _require(match.rank > 0 and match.score > 0, "rank or score is not positive")
        cases.append(ConformanceCase("record_preservation", True))
    except Exception as error:
        cases.append(_failed("record_preservation", error))

    try:
        isolated = await provider.search(
            _search_request(run_id, "isolated-search", other_subject, "solarized editor preference")
        )
        _require(
            any(item.record.id == stored.record.id for item in searched.matches),
            "subject search did not cross sessions",
        )
        _require(not isolated.matches, "search crossed the subject boundary")
        cases.append(ConformanceCase("subject_isolation_cross_session", True))
    except Exception as error:
        cases.append(_failed("subject_isolation_cross_session", error))

    try:
        first = await provider.search(search_request)
        second = await provider.search(search_request)
        _require(first == second, "repeated search changed results or order")
        cases.append(ConformanceCase("deterministic_order", True))
    except Exception as error:
        cases.append(_failed("deterministic_order", error))

    try:
        replay = await provider.store(
            MemoryStoreRequest(
                context=_context(run_id, "replay"),
                namespace=stored_request.namespace,
                content=stored_request.content,
                event_timestamp=stored_request.event_timestamp,
                provenance=stored_request.provenance,
                metadata=stored_request.metadata,
                idempotency_key=stored_request.idempotency_key,
            )
        )
        _require(replay.disposition is MemoryStoreDisposition.EXISTING, "replay did not report existing")
        _require(replay.record == stored.record, "replay did not return original record")
        conflict_request = MemoryStoreRequest(
            context=_context(run_id, "conflict"),
            namespace=stored_request.namespace,
            content=MemoryContent.text_content("different conformance payload"),
            event_timestamp=stored_request.event_timestamp,
            provenance=stored_request.provenance,
            metadata=stored_request.metadata,
            idempotency_key=stored_request.idempotency_key,
        )
        try:
            await provider.store(conflict_request)
        except MemoryProviderError as error:
            _require(error.error.code is MemoryErrorCode.CONFLICT, "conflict returned the wrong code")
        else:
            raise AssertionError("conflicting replay unexpectedly succeeded")
        cases.append(ConformanceCase("idempotency", True))
    except Exception as error:
        cases.append(_failed("idempotency", error))

    try:
        filtered = await provider.search(
            MemorySearchRequest(
                context=_context(run_id, "filtered"),
                namespace=cross_session,
                query="solarized editor preference",
                filter=MemoryFilter(
                    metadata={"conformance_kind": "preference"},
                    event_after=stored.record.event_timestamp - timedelta(seconds=1),
                    event_before=stored.record.event_timestamp + timedelta(seconds=1),
                    ingested_before=stored.record.ingested_at + timedelta(seconds=1),
                ),
                limit=4,
            )
        )
        rejected = await provider.search(
            MemorySearchRequest(
                context=_context(run_id, "filtered-empty"),
                namespace=cross_session,
                query="solarized editor preference",
                filter=MemoryFilter(metadata={"conformance_kind": "other"}),
                limit=4,
            )
        )
        _require(any(item.record.id == stored.record.id for item in filtered.matches), "matching filter lost record")
        _require(not rejected.matches, "nonmatching metadata filter returned records")
        cases.append(ConformanceCase("metadata_time_filters", True))
    except Exception as error:
        cases.append(_failed("metadata_time_filters", error))

    try:
        invalid = _search_request(run_id, "invalid", cross_session, " ")
        try:
            await provider.search(invalid)
        except MemoryContractError as error:
            _require(error.error.code is MemoryErrorCode.INVALID_REQUEST, "invalid request returned wrong code")
        except MemoryProviderError as error:
            _require(error.error.code is MemoryErrorCode.INVALID_REQUEST, "invalid request returned wrong code")
        else:
            raise AssertionError("invalid search unexpectedly succeeded")
        cases.append(ConformanceCase("typed_invalid_request", True))
    except Exception as error:
        cases.append(_failed("typed_invalid_request", error))

    try:
        await _check_capabilities(provider, run_id, origin)
        cases.append(ConformanceCase("capability_agreement", True))
    except Exception as error:
        cases.append(_failed("capability_agreement", error))

    return ConformanceReport(provider.name, run_id, tuple(cases))


async def _check_capabilities(provider: MemoryProvider, run_id: str, namespace: MemoryNamespace) -> None:
    capabilities: MemoryCapabilities = provider.capabilities
    unsupported_python_profile = {
        "update": capabilities.update,
        "delete": capabilities.delete,
        "batch_store": capabilities.batch_store,
        "feedback": capabilities.feedback,
        "health": capabilities.health,
    }
    enabled = [name for name, advertised in unsupported_python_profile.items() if advertised]
    _require(not enabled, f"Python adapter profile cannot verify advertised capabilities: {', '.join(enabled)}")
    maintain = getattr(provider, "maintain", None)
    if not capabilities.maintenance:
        return
    if not callable(maintain):
        raise AssertionError("maintenance is advertised but maintain is absent")
    request = MemoryMaintenanceRequest(
        context=_context(run_id, "maintenance"),
        namespace=namespace,
        action=MemoryMaintenanceAction.REFLECT,
        window=MemoryMaintenanceWindow(
            checkpoint_id=f"{run_id}-checkpoint",
            query="Reflect on the conformance memory",
            limit=5,
        ),
    )
    result = await maintain(request)
    _require(isinstance(result, MemoryMaintenanceResult), "maintain returned the wrong result type")


def _namespace(run_id: str, subject: str, session: str, agent: str) -> MemoryNamespace:
    return MemoryNamespace(f"relay-conformance-{run_id}", subject, session, agent)


def _context(run_id: str, suffix: str) -> MemoryRequestContext:
    return MemoryRequestContext(f"{run_id}-{suffix}")


def _store_request(
    run_id: str,
    suffix: str,
    namespace: MemoryNamespace,
    text: str,
    idempotency_key: str,
) -> MemoryStoreRequest:
    return MemoryStoreRequest(
        context=_context(run_id, suffix),
        namespace=namespace,
        content=MemoryContent.text_content(text),
        event_timestamp=datetime(2026, 6, 29, 12, tzinfo=timezone.utc),
        provenance=MemoryProvenance("conformance", (f"{run_id}-turn",)),
        metadata={"conformance_kind": "preference"},
        idempotency_key=f"{run_id}-{idempotency_key}",
    )


def _search_request(
    run_id: str,
    suffix: str,
    namespace: MemoryNamespace,
    query: str,
) -> MemorySearchRequest:
    return MemorySearchRequest(
        context=_context(run_id, suffix),
        namespace=namespace,
        query=query,
        scope=MemorySearchScope.SUBJECT,
        limit=10,
    )


def _require(condition: object, message: str) -> None:
    if not condition:
        raise AssertionError(message)


def _failed(name: str, error: Exception) -> ConformanceCase:
    return ConformanceCase(name, False, f"{type(error).__name__}: {error}")


__all__ = ["ConformanceCase", "ConformanceReport", "run_provider_conformance"]

File: crates/memory/src/conformance.rs
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

//! Reusable provider conformance coverage.

use std::collections::BTreeMap;
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
use std::time::{Duration, SystemTime};

use async_trait::async_trait;
use nemo_relay_types::memory::{
    MemoryBatchStoreRequest, MemoryContent, MemoryDeleteRequest, MemoryErrorCode,
    MemoryFeedbackKind, MemoryFeedbackRequest, MemoryFilter, MemoryHealthRequest,
    MemoryMaintenanceAction, MemoryMaintenanceRequest, MemoryNamespace, MemoryOperationError,
    MemoryProvenance, MemoryRequestContext, MemorySearchRequest, MemorySearchResult,
    MemorySearchScope, MemoryStoreDisposition, MemoryStoreRequest, MemoryStoreResult,
    MemoryUpdateRequest,
};

use crate::provider::{MemoryProvider, MemoryProviderResult};
use crate::runtime::MemoryRuntime;

/// Result of one named provider conformance case.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ConformanceCase {
    /// Stable case name suitable for test and CI output.
    pub name: &'static str,
    /// Whether every assertion in the case passed.
    pub passed: bool,
    /// Diagnostic summary when the case failed.
    pub message: Option<String>,
}

/// Structured result of one provider conformance run.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ConformanceReport {
    /// Provider identifier reported by the implementation.
    pub provider: String,
    /// Caller-supplied identifier used to isolate synthetic test data.
    pub run_id: String,
    /// Named cases in execution order.
    pub cases: Vec<ConformanceCase>,
}

impl ConformanceReport {
    /// Return `true` when every conformance case passed.
    pub fn passed(&self) -> bool {
        self.cases.iter().all(|case| case.passed)
    }

    /// Return only failed cases.
    pub fn failures(&self) -> impl Iterator<Item = &ConformanceCase> {
        self.cases.iter().filter(|case| !case.passed)
    }

    fn record(&mut self, name: &'static str, result: Result<(), String>) {
        match result {
            Ok(()) => self.cases.push(ConformanceCase {
                name,
                passed: true,
                message: None,
            }),
            Err(message) => self.cases.push(ConformanceCase {
                name,
                passed: false,
                message: Some(message),
            }),
        }
    }
}

/// Run provider-neutral acceptance cases against one isolated provider namespace.
///
/// `run_id` must be unique for each invocation against a durable or shared
/// backend. The harness reports failures as named data instead of panicking.
/// It tests required store/search behavior, record preservation, namespace
/// isolation, deterministic order, idempotency, optional capability agreement,
/// and direct-runtime deadline/partial-result invariants.
pub async fn run_provider_conformance(
    provider: Arc<dyn MemoryProvider>,
    run_id: impl Into<String>,
) -> ConformanceReport {
    let run_id = run_id.into();
    let mut report = ConformanceReport {
        provider: provider.name().to_string(),
        run_id: run_id.clone(),
        cases: vec![],
    };
    if run_id.trim().is_empty() {
        report.record("valid_run_id", Err("run_id must not be empty".to_string()));
        return report;
    }

    let runtime = MemoryRuntime::from_arc(provider);
    let origin = namespace(&run_id, "subject", "session-a", "assistant");
    let cross_session = namespace(&run_id, "subject", "session-b", "assistant");
    let stored_request = store_request(
        &run_id,
        "required-store",
        origin.clone(),
        "conformance solarized editor preference",
        Some("required-store-key"),
    );
    let stored = runtime.store(stored_request.clone()).await;
    report.record(
        "required_store",
        stored
            .as_ref()
            .map(|result| {
                require(
                    result.disposition == MemoryStoreDisposition::Created,
                    "initial store did not report created",
                )
            })
            .unwrap_or_else(|error| Err(error.to_string())),
    );

    let Some(stored) = stored.ok() else {
        for name in [
            "required_search",
            "record_preservation",
            "subject_isolation_cross_session",
            "deterministic_order",
            "idempotency",
            "capability_agreement",
        ] {
            report.record(name, Err("required seed store failed".to_string()));
        }
        record_runtime_invariants(&mut report).await;
        return report;
    };

    let search = runtime
        .search(search_request(
            &run_id,
            "required-search",
            cross_session.clone(),
            "solarized editor preference",
        ))
        .await;
    report.record(
        "required_search",
        search
            .as_ref()
            .map_err(ToString::to_string)
            .and_then(|result| require(!result.matches.is_empty(), "search returned no matches")),
    );
    report.record(
        "record_preservation",
        search
            .as_ref()
            .map_err(ToString::to_string)
            .and_then(|result| {
                result
                    .matches
                    .iter()
                    .find(|memory_match| memory_match.record.id == stored.record.id)
                    .ok_or_else(|| "stored record was not returned".to_string())
                    .and_then(|memory_match| {
                        require(
                            memory_match.record == stored.record
                                && memory_match.rank > 0
                                && memory_match.score > 0.0,
                            "record fields, rank, or score changed",
                        )
                    })
            }),
    );

    let isolated = runtime
        .search(search_request(
            &run_id,
            "isolated-search",
            namespace(&run_id, "other-subject", "session-b", "assistant"),
            "solarized editor preference",
        ))
        .await;
    report.record(
        "subject_isolation_cross_session",
        search
            .as_ref()
            .map_err(ToString::to_string)
            .and_then(|result| {
                require(
                    result
                        .matches
                        .iter()
                        .any(|memory_match| memory_match.record.id == stored.record.id),
                    "subject-scope search did not cross sessions",
                )
            })
            .and_then(|()| {
                isolated
                    .as_ref()
                    .map_err(ToString::to_string)
                    .and_then(|result| {
                        require(result.matches.is_empty(), "search crossed subject boundary")
                    })
            }),
    );

    let deterministic_request = search_request(
        &run_id,
        "deterministic-search",
        cross_session,
        "solarized editor preference",
    );
    let deterministic_first = runtime.search(deterministic_request.clone()).await;
    let deterministic_second = runtime.search(deterministic_request).await;
    report.record(
        "deterministic_order",
        match (deterministic_first, deterministic_second) {
            (Ok(first), Ok(second)) => require(first == second, "repeated search changed order"),
            (Err(error), _) | (_, Err(error)) => Err(error.to_string()),
        },
    );

    let mut replay_request = stored_request.clone();
    replay_request.context = context(&run_id, "idempotent-replay");
    let replay = runtime.store(replay_request).await;
    let mut conflict_request = stored_request;
    conflict_request.context = context(&run_id, "idempotent-conflict");
    conflict_request.content = MemoryContent::Text {
        text: "different conformance payload".to_string(),
    };
    let conflict = runtime.store(conflict_request).await;
    report.record(
        "idempotency",
        replay
            .as_ref()
            .map_err(ToString::to_string)
            .and_then(|result| {
                require(
                    result.disposition == MemoryStoreDisposition::Existing
                        && result.record == stored.record,
                    "idempotent replay did not return the original record",
                )
            })
            .and_then(|()| match conflict {
                Err(error) if error.code == MemoryErrorCode::Conflict => Ok(()),
                Err(error) => Err(format!("expected conflict, got {:?}", error.code)),
                Ok(_) => Err("conflicting replay unexpectedly succeeded".to_string()),
            }),
    );

    report.record(
        "capability_agreement",
        check_capability_agreement(&runtime, &run_id, origin, &stored.record.id).await,
    );
    record_runtime_invariants(&mut report).await;
    report
}

async fn check_capability_agreement(
    runtime: &MemoryRuntime,
    run_id: &str,
    namespace: MemoryNamespace,
    record_id: &str,
) -> Result<(), String> {
    let capabilities = runtime.capabilities();
    let update = runtime
        .update(MemoryUpdateRequest {
            context: context(run_id, "capability-update"),
            namespace: namespace.clone(),
            id: record_id.to_string(),
            content: None,
            metadata: Some(BTreeMap::new()),
        })
        .await
        .map(|_| ());
    agreement("update", capabilities.update, update)?;

    let batch = runtime
        .batch_store(MemoryBatchStoreRequest {
            context: context(run_id, "capability-batch"),
            requests: vec![store_request(
                run_id,
                "capability-batch-item",
                namespace.clone(),
                "conformance batch item",
                Some("capability-batch-key"),
            )],
        })
        .await
        .map(|_| ());
    agreement("batch_store", capabilities.batch_store, batch)?;

    let maintenance = runtime
        .maintain(MemoryMaintenanceRequest {
            context: context(run_id, "capability-maintenance"),
            namespace: namespace.clone(),
            action: MemoryMaintenanceAction::Reflect,
            parameters: BTreeMap::new(),
        })
        .await
        .map(|_| ());
    agreement("maintenance", capabilities.maintenance, maintenance)?;

    let feedback = runtime
        .feedback(MemoryFeedbackRequest {
            context: context(run_id, "capability-feedback"),
            namespace: namespace.clone(),
            id: record_id.to_string(),
            kind: MemoryFeedbackKind::Relevant,
            confidence: Some(1.0),
            metadata: BTreeMap::new(),
        })
        .await
        .map(|_| ());
    agreement("feedback", capabilities.feedback, feedback)?;

    let health = runtime
        .health(MemoryHealthRequest {
            context: context(run_id, "capability-health"),
        })
        .await
        .map(|_| ());
    agreement("health", capabilities.health, health)?;

    let delete = runtime
        .delete(MemoryDeleteRequest {
            context: context(run_id, "capability-delete"),
            namespace,
            id: record_id.to_string(),
        })
        .await
        .map(|_| ());
    agreement("delete", capabilities.delete, delete)
}

fn agreement(
    capability: &str,
    advertised: bool,
    result: Result<(), MemoryOperationError>,
) -> Result<(), String> {
    match (advertised, result) {
        (false, Err(error)) if error.code == MemoryErrorCode::Unsupported => Ok(()),
        (false, Err(error)) => Err(format!(
            "{capability} is disabled but returned {:?}",
            error.code
        )),
        (false, Ok(())) => Err(format!("{capability} is disabled but succeeded")),
        (true, Err(error)) if error.code == MemoryErrorCode::Unsupported => Err(format!(
            "{capability} is advertised but returned unsupported"
        )),
        (true, Err(_) | Ok(())) => Ok(()),
    }
}

async fn record_runtime_invariants(report: &mut ConformanceReport) {
    let partial_error = MemoryOperationError::new(
        MemoryErrorCode::ProviderUnavailable,
        "synthetic partial failure",
        true,
    );
    let partial_result = MemorySearchResult {
        matches: vec![],
        partial_errors: vec![partial_error],
    };
    let partial_runtime = MemoryRuntime::new(RuntimeProbe {
        response: partial_result.clone(),
        block: false,
        calls: Arc::new(AtomicUsize::new(0)),
        dropped: Arc::new(AtomicBool::new(false)),
    });
    report.record(
        "partial_result_preservation",
        partial_runtime
            .search(search_request(
                "runtime",
                "partial",
                namespace("runtime", "subject", "session", "agent"),
                "synthetic query",
            ))
            .await
            .map_err(|error| error.to_string())
            .and_then(|actual| require(actual == partial_result, "partial result changed")),
    );

    let calls = Arc::new(AtomicUsize::new(0));
    let dropped = Arc::new(AtomicBool::new(false));
    let deadline_runtime = MemoryRuntime::new(RuntimeProbe {
        response: MemorySearchResult::default(),
        block: true,
        calls: calls.clone(),
        dropped: dropped.clone(),
    });
    let mut deadline_request = search_request(
        "runtime",
        "deadline",
        namespace("runtime", "subject", "session", "agent"),
        "synthetic query",
    );
    deadline_request.context.deadline =
        Some((SystemTime::now() + Duration::from_millis(50)).into());
    let deadline_result = deadline_runtime.search(deadline_request).await;
    report.record(
        "deadline_cancellation",
        match deadline_result {
            Err(error) => require(
                error.code == MemoryErrorCode::DeadlineExceeded
                    && calls.load(Ordering::SeqCst) == 1
                    && dropped.load(Ordering::SeqCst),
                "deadline did not drop exactly one provider future",
            ),
            Ok(_) => Err("blocking provider completed before deadline".to_string()),
        },
    );
}

struct RuntimeProbe {
    response: MemorySearchResult,
    block: bool,
    calls: Arc<AtomicUsize>,
    dropped: Arc<AtomicBool>,
}

#[async_trait]
impl MemoryProvider for RuntimeProbe {
    fn name(&self) -> &str {
        "conformance_runtime_probe"
    }

    async fn search(
        &self,
        _request: MemorySearchRequest,
    ) -> MemoryProviderResult<MemorySearchResult> {
        self.calls.fetch_add(1, Ordering::SeqCst);
        if self.block {
            let _guard = DropFlag(self.dropped.clone());
            std::future::pending().await
        } else {
            Ok(self.response.clone())
        }
    }

    async fn store(&self, _request: MemoryStoreRequest) -> MemoryProviderResult<MemoryStoreResult> {
        Err(MemoryOperationError::new(
            MemoryErrorCode::Internal,
            "runtime probe does not store",
            false,
        ))
    }
}

struct DropFlag(Arc<AtomicBool>);

impl Drop for DropFlag {
    fn drop(&mut self) {
        self.0.store(true, Ordering::SeqCst);
    }
}

fn namespace(run_id: &str, subject: &str, session: &str, agent: &str) -> MemoryNamespace {
    MemoryNamespace {
        tenant_id: format!("conformance-{run_id}"),
        subject_id: subject.to_string(),
        session_id: Some(session.to_string()),
        agent_id: Some(agent.to_string()),
    }
}

fn context(run_id: &str, operation: &str) -> MemoryRequestContext {
    MemoryRequestContext {
        operation_id: format!("{run_id}-{operation}"),
        deadline: None,
    }
}

fn store_request(
    run_id: &str,
    operation: &str,
    namespace: MemoryNamespace,
    text: &str,
    idempotency_key: Option<&str>,
) -> MemoryStoreRequest {
    MemoryStoreRequest {
        context: context(run_id, operation),
        namespace,
        content: MemoryContent::Text {
            text: text.to_string(),
        },
        event_timestamp: (SystemTime::UNIX_EPOCH + Duration::from_secs(1_700_000_000)).into(),
        provenance: MemoryProvenance {
            source: "conformance".to_string(),
            source_ids: vec![format!("source-{run_id}-{operation}")],
            ..MemoryProvenance::default()
        },
        metadata: BTreeMap::from([("suite".to_string(), "memory".into())]),
        idempotency_key: idempotency_key.map(|key| format!("{run_id}-{key}")),
    }
}

fn search_request(
    run_id: &str,
    operation: &str,
    namespace: MemoryNamespace,
    query: &str,
) -> MemorySearchRequest {
    MemorySearchRequest {
        context: context(run_id, operation),
        namespace,
        query: query.to_string(),
        scope: MemorySearchScope::Subject,
        filter: MemoryFilter::default(),
        limit: 10,
    }
}

fn require(condition: bool, message: &str) -> Result<(), String> {
    condition.then_some(()).ok_or_else(|| message.to_string())
}
